### PR TITLE
consensus_encoding: re-org tests to prefer integration over inline

### PR DESCRIPTION
**Pull request by nyonson**
This is currently (mostly) a big dumb move of existing tests, but hoping to get some feedback on the general direction. I am proposing that we prefer integration tests over inline ones since they are against the exposed API. In other words, a test lives in an integration if and only if every name it touches is reachable through `lib.rs`. I think it's generally better to test against the API so we don't do something under the hood by accident. This also defines a default spot to drop new tests instead of spreading them across integration and inline. Hoping this makes it easier to eyeball if there are some gaps in the tests.

This moves all existing inline tests which do not depend on accessing private items over to integration.

If we are cool with this principle for tests, I'd like to follow this up by cleaning up the integration tests, maybe find a better way to organize them, add some more edge cases, and more round trip tests.

### CODE REVIEW

File: consensus_encoding/tests/decode.rs [hunk 1]
```diff
@@ -1,13 +1,151 @@
 // SPDX-License-Identifier: CC0-1.0
 
-//! Integration tests for decode module.
+//! Decoding tests.
 
 use bitcoin_consensus_encoding::{
```
Comment (tcharding) [inline consensus_encoding/tests/decode.rs]:
```suggestion
use bitcoin_consensus_encoding::{
    self as encoding,
```

Then use `encoding::` everywhere.

Comment (nyonson) [reply consensus_encoding/tests/decode.rs]: Ugh, yea I got lazy here, this copy/pasta move took me way longer than I want to admit...

File: consensus_encoding/tests/encode.rs [hunk 2]
```diff
@@ -219,3 +242,438 @@ fn encode_option_encoder_none() {
...
+
+    assert_eq!(encoder.current_chunk(), &[0x34, 0x12, 0x00, 0x00][..]);
+    assert!(encoder.advance());
+    assert_eq!(encoder.current_chunk(), &[0x78, 0x56, 0x00, 0x00][..]);
+    assert!(!encoder.advance());
+    assert!(encoder.current_chunk().is_empty());
+}
+
+#[test]
+#[cfg(feature = "alloc")]
```
Comment (tcharding) [inline consensus_encoding/tests/encode.rs]: This doesn't need `alloc`.

Comment (tcharding) [reply consensus_encoding/tests/encode.rs]: Same suggestion in a bunch of other places too mate.

**Comment by tcharding** [comment]
This seems reasonable to me. Also FTR its easy enough to review because its mostly code moves.

File: consensus_encoding/tests/decode.rs [hunk 3]
```diff
@@ -217,3 +245,136 @@ fn decode_byte_vec_decoder_does_not_overconsume_on_second_chunk() {
...
+    let result: Result<TestArray, _> = decode_from_slice(&data);
+    assert!(result.is_ok());
+    let decoded = result.unwrap();
+    assert_eq!(decoded.0, [1, 2, 3, 4]);
+}
+
+#[test]
+#[cfg(feature = "std")]
+fn decode_from_read_extra_data() {
+    use std::io::Cursor;
```
Comment (mpbagot) [inline consensus_encoding/tests/decode.rs]: Similar to Tobin's comment, you could move this to a feature gated `use` at the top and drop all of them from each test below

**Comment by nyonson** [comment]
7f4e8320268691f7c5e01ba29ce0e028fef8a749: rebased and folded in @mpbagot and @tcharding's clean up suggestions. These were pretty mechanical so doesn't change the 1:1 test move aspect of the change.

I also went on a little journey experimenting with preferring an inline `tests` module in the root `lib.rs` instead of the integration `tests` directory, since the visibility rules would be about the same. I think we should stick with this integration layout, but if these subtleties interest you (hey, there is a chance!), I [wrote up the journey](https://blog.yonson.dev/log/2026-02-vol4/).

**Comment by nyonson** [comment]
b8bc8ac274d0a90970d44a055341d7563df2ed16: fix formatting

**Comment by tcharding** [comment]
This got waay harder to review. Yesterday it was almost all code moves, today there is loads of red/green line changes.

File: consensus_encoding/tests/encode.rs [hunk 4]
```diff
@@ -1,44 +1,64 @@
...
+#[cfg(any(feature = "alloc", feature = "std"))]
 struct TestData(u32);
 
-#[cfg(feature = "alloc")]
-impl Encodable for TestData {
+#[cfg(any(feature = "alloc", feature = "std"))]
+impl encoding::Encodable for TestData {
     type Encoder<'e>
-        = ArrayEncoder<4>
+        = encoding::ArrayEncoder<4>
```
Comment (tcharding) [inline consensus_encoding/tests/encode.rs]: Bother, I guess you are trying to follow the style introduced in https://github.com/rust-bitcoin/rust-bitcoin/pull/5583? Can this be done as an initial patch so as to reduce the diff of later patches? Also there are formatting changes in the first patch.

**Comment by nyonson** [comment]
> This got waay harder to review. Yesterday it was almost all code moves, today there is loads of red/green line changes. 

Dang I think I misinterpreted your initial review comment and also ran a format. I can tease the commits apart though.

**Comment by tcharding** [comment]
Ah sorry if I caused that.

**Comment by nyonson** [comment]
c4143bb6f08d15cd3ae6f3b9ffd940ecba3ead37: revert to big dumb move for ease of reviewing. If this pattern is adopted, will follow up with some cleanup patches.

**Comment by tcharding** [comment]
Thanks man.
